provisio-core/src/main/java/ca/vanzyl/provisio/action/artifact/WriteToDiskAction.java

```diff
@@ -41,20 +49,37 @@ public WriteToDiskAction(ProvisioArtifact artifact, File outputDirectory) {
 
     @Override
     public void execute(ProvisioningContext context) {
-        File file = artifact.getFile();
-        if (file != null) {
-            copy(file, new File(outputDirectory, targetArtifactFileName(artifact, file.getName())));
+        if (artifact.getFile() != null) {
+            write(
+                    context,
+                    artifact,
+                    targetArtifactFileName(context, artifact, artifact.getFile().getName()));
         }
     }
 
-    public void copy(File source, File target) {
+    private void write(ProvisioningContext context, ProvisioArtifact source, String targetPath) {
+        File target = new File(outputDirectory, targetPath);
         try {
-            if (!target.getParentFile().exists()) {
-                target.getParentFile().mkdirs();
+            Files.createDirectories(target.getParentFile().toPath());
+            if (Files.exists(target.toPath())) {
+                if (ProvisioVariables.allowTargetOverwrite(context)) {
+                    logger.warn("Conflict: artifact {} overwrites existing file {}", artifact, targetPath);
+                } else {
+                    throw new ProvisioningException(
+                            "Conflict: artifact " + artifact + " would overwrite existing file: " + targetPath);
+                }
+            }
+            try (CachingOutputStream outputStream = new CachingOutputStream(target.toPath(), BUFFER_SIZE);
```

Comment:
Use explicit buffer size is not needed, as below transferTo will again use default (8k) buffer.